### ISSUE TITLE
Optimize Top 5 Movies

### ISSUE THREAD

**Issue by DevMan01** [Casvt/Plex-scripts#147]
Hey there, I need help with a script that does the following:

- Automatically optimize the top 5movies. I'd like this to be a rolling list, so, as something becomes the top 5, Plex will automatically queue that movie for optimization. Then likewise, if something falls off the top 5, the optimized version will be deleted.

- I do use Plex Meta Manager, as well as Tautulli. So, its could be easy for me to use Plex Meta Manager to create a collection of the top 5 movies, using analytics from Tautulli. 

- A Python, or .sh framework would work.

**Comment by splitsec2** [comment]
A slight variation on this would be to have the option to optimize things in the watchlist. Which in theory are likely to be watched,

**Comment by DevMan01** [comment]
I like that. Perhaps the best solution would be to parameratize the script in such a way that it could receive any playlist, watchlist or collection, and optimize as needed. 

I’ve also been wondering / brainstorming how the script could keep track of changes in the list/collection. That way when an item leaves the list, it will delete the optimized version.

**Comment by Casvt** [comment]
> I’ve also been wondering / brainstorming how the script could keep track of changes in the list/collection. That way when an item leaves the list, it will delete the optimized version.

The script is constantly running in the background. It fetches the content of the list and compares it to the last time it fetched. Items that are in the list but aren't in the previous list are newly added and will be triggered to be optimized. Items that don't appear in the new list but do in the previous are removed and thus will have their optimized version removed.

**Comment by DevMan01** [comment]
Agreed. The question in my mind is how we can do the diff between the two states (current versus previous). Do we write it in a state file of some kind? Perhaps a JSON list containing unique id’s per list? It is it all kept in memory?

**Comment by Casvt** [comment]
The script is constantly running, like a service. We store the previous fetch in a variable.

**Comment by splitsec2** [comment]
If there is no persistent storage what happens during a reboot? I guess the window is small for the race condition of the script grabbing the “current” and plex scanning the library and discovering a change or the user changing the watchlist

**Comment by Casvt** [comment]
The first time the script fetches, when there is no previous fetch available, the script goes through every media entry in the list and checks if it's already optimized.

From then on we don't have to do this everytime and can decide based on the current and previous fetch and the differences between them.
